### DISCUSSION TITLE
Connect more than 8x 18B20 Temp.-Sensors

### DISCUSSION THREAD

**Discussion by DasMoritz** [arendst/Tasmota#13943] [Q&A] [answered]
Hello together, 

I am using an ESP32 with eight 18B20 temperature-senors connected via one-wire to GPIO4. 
Is it possible to use more than eight sensors via onewire on the same GPIO Port? 

If I have understood it correctly it is not possible to connect eight sensors on GPIO4 and further / additional sensors on a second GPIO-Port (for example GPIO5). 

What do I want to do: I would like to controll and monitor my heating system with 6 heating-circuits and two pipes for each circuit so that I need 12 temp.-sensors. 

Thank you very much and sorry for my bad english. 
Moritz

**Comment by sfromis** [comment, marked as answer]
As [documented already](https://tasmota.github.io/docs/DS18x20/#multiple-sensors), Tasmota allows more than 8 DS18B20 sensors on the same one-wire bus, if you [build](https://tasmota.github.io/docs/Compile-your-build/) with:
`#define DS18X20_MAX_SENSORS 12`

**Reply by DasMoritz** [reply]
Hi, thank you first of all. 

I get an suceed message in Visual Studio Code as you can see: 
![image](https://user-images.githubusercontent.com/73671970/145061852-ba653666-2263-4e2b-8651-1b653bd211e8.png)

If I flash the ESP32 with ESPflasher (based on esp pyflasher) I will also get an suceed but after that.... 
![image](https://user-images.githubusercontent.com/73671970/145062063-51c4188a-d8d9-4694-9e8e-dcdd8aa452a6.png)
(the lower part is shown several times per second).

Any idea?

**Reply by sfromis** [reply]
For ESP32, you need to build tasmota32 instead of just "tasmota", the latter is for ESP8266. IIRC, you can select it near the top, just below "PIO Home".

**Reply by DasMoritz** [reply]
Hm, I don't find any "tasmota32" in the VSC. 
Is it necessary to build a custom tasmota.bin for the ESP32 or does the tasmota32.bin support more than 8 sensors per default?

**Reply by barbudor** [reply]
Search better
If you click on the bottom line on *default* you will see the list of available environments including tasmota32
You still have to override the default value of DS18X20_MAX_SENSORS in your user_config_override.h

**Comment by DasMoritz** [comment]
Hi, 

ok, now I have selected ESP-32 in the blue bottom line: 
![image](https://user-images.githubusercontent.com/73671970/145099189-284bfb54-620c-4962-a46e-0e1b4d302f72.png)

After that I changed the xsns_05_ds18x20.ino (L42) to 12 and added "# define DS18X20_MAX_SENSORS 12" (without "") to the user_config_override.h.

Next Step: Terminal --> Run Build Task --> Grab a coffee --> And than (as you can see on the screenshot) I have created a Firmware.bin with 162KB, that seems to bit low. 

What I am also doing wrong?

**Reply by barbudor** [reply]
You still haven't select `tasmota32` instead of Default
By clicking on "Default" you should have this list dropping from the top of the screen:
![image](https://user-images.githubusercontent.com/2996042/145112988-3c091133-1ed6-415d-b795-4dfb00404b45.png)
...
![image](https://user-images.githubusercontent.com/2996042/145113050-5616843f-f83c-4e83-9ffb-6708c08b6788.png)

Then you will see:
![image](https://user-images.githubusercontent.com/2996042/145111167-ca3d45f1-ed45-41f3-9412-e5639a0a79f0.png)

Then click on the tick mark:
![image](https://user-images.githubusercontent.com/2996042/145111504-14048e16-fdab-421f-b79d-ca28fdaae054.png)

The log in your screen shot are very strange
Nowhere you should see a reference to "az-delivery-devkit-v4"

My build log starts with:
```
> Executing task in folder Tasmota: C:\Users\Jean-Michel\.platformio\penv\Scripts\platformio.exe run --environment tasmota32 <

Processing tasmota32 (framework: arduino; platform: https://github.com/tasmota/platform-espressif32/releases/download/v3.4.1/Tasmota-platform-espressif32.zip; board: esp32_4M)
----------------------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32_4M.html
PLATFORM: Espressif 32 (3.4.1) > Espressif Generic ESP32 4M Flash, Tasmota 1856k Code/OTA, 320k SPIFFS
HARDWARE: ESP32 80MHz, 320KB RAM, 4MB Flash
DEBUG: Current (esp-prog) External (esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES:
 - framework-arduinoespressif32 2.0.1+1
 - tool-esptoolpy 1.30200.211025 (3.2.0)
 - tool-mklittlefs 1.203.210628 (2.3)
 - toolchain-xtensa-esp32 8.4.0+2021r2
Converting tasmota.ino
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ strict
Found 112 compatible libraries
Scanning dependencies...`
```
and ends with:
```
Linking .pio\build\tasmota32\firmware.elf
Retrieving maximum program size .pio\build\tasmota32\firmware.elf
Checking size .pio\build\tasmota32\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [==        ]  18.2% (used 59532 bytes from 327680 bytes)
Flash: [=======   ]  70.6% (used 1341161 bytes from 1900544 bytes)
Building .pio\build\tasmota32\firmware.bin
esptool.py v3.2
Merged 25 ELF sections
bin_map_copy([".pio\build\tasmota32\firmware.bin"], [".pio\build\tasmota32\firmware.elf"])
map_gzip([".pio\build\tasmota32\firmware.bin"], [".pio\build\tasmota32\firmware.elf"])
==================================================== [SUCCESS] Took 964.45 seconds ====================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota32      SUCCESS   00:16:04.452
===================================================== 1 succeeded in 00:16:04.452 =====================================================

Terminal will be reused by tasks, press any key to close it.
```

**Reply by DasMoritz** [reply]
Hi @barbudor and @sfromis 

now it is working perfect :-) 
Thank you very much and sorry for the inconvenience.

**Comment by Bender2k** [comment]
Hi, i have to ask for help regarding this topic and hope someone still reads this discussion.

I installed 12.2 (Web Installer) on an ESP8266 and connected 10 DS18B20 Sensors, but can of course only see 8.
In another discussion i read that using a second GPIO pin does not work, right? Or does it? (The post was relatively old)

The problem is, i am just a hobbyist and struggle using Arduino IDE and all these librarys. I have seen the .ino file of Tasmota includes 20-30 dependencies i don't have. Solving all these issues is something i cant manage probably. :(

Is there another way how i can use 10 sensors on one ESP? Why is this variable not configurable in the Tasmota WebInterface or at least with an console command? (Is it?)

Thank you!

**Reply by sfromis** [reply]
With Tasmota you can easily support 10 DS18B20 sensors on one pin, but you need a custom build to go higher than the default of 8. There are web-based online [tools to do a custom build](https://tasmota.github.io/docs/Compile-your-build/#compiling-tools), you need not worry about Arduino IDE and libraries. It is pretty easy to do.

**Reply by barbudor** [reply]
You can't compile Tasmota with Arduino IDE
It requires platformio, generally used with VSCode

If you are not at ease with development environment, you can have a look at TasmoCompiler, a web based tool for building your own binaries

For more than 8 DS18X20, you need to compile tasmota with the following line in your user_config_override.h (or the last box in Tasmocompiler) :
```
#define DS18X20_MAX_SENSORS 10
```
That would allow up to 10 sensors on a single GPIO
A recent feature allows to dispatch the 10 DS on up to 4 GPIOs

**Reply by Jason2866** [reply]
The goal of the Platformio Tasmota setup is not to fight to get all needed stuff together.
So we decided (it is been a while now) to drop ArduinoIDE support since it is horrible to use with complex projects.
Using Platformio to compile Tasmota, is really simple and described in Tasmota documentation.
You dont have (and you shouldnt!) install anything else than VSC/Platformio and simpy download (or git clone) Tasmota source in a folder. Thats all!

**Reply by Bender2k** [reply]
> With Tasmota you can easily support 10 DS18B20 sensors on one pin, but you need a custom build to go higher than the default of 8. There are web-based online [tools to do a custom build](https://tasmota.github.io/docs/Compile-your-build/#compiling-tools), you need not worry about Arduino IDE and libraries. It is pretty easy to do.

SHIT THAT WAS EASY !!!!

10 Sensors now working. Thank you!

**Reply by Bender2k** [reply]
Haha funny... Just after everything worked i have seen the Tasmota 12.2.0.1 changelog:

**DS18x20 support on up to four GPIOs**

maybe interesting for others who find this dicussion. :)

**Reply by sfromis** [reply]
You still need to config max number to have more than 8.

**Reply by spcqike** [reply]
@sfromis if one still need to increase the max number manually, what is the advantage of using multiple GPIOs in the first place? 

I thought with 4 GPIOs and 8 sensors max, one could utilize 32 sensors. 

And out of curiosity: does anyone know if there is a maximum limit for the amount of sensors? In theory 1-wire supports like 100 slaves? But is this really possible with an esp? Or with Tasmota? :)

**Reply by sfromis** [reply]
Advantages of spreading across multiple pins could be flexibility/reliability in wiring and changes, easier use of hardware already designed for more than one one-wire pin, and maybe easier management of the need for suitable pullup.

About number of sensors, do read the docs: https://tasmota.github.io/docs/DS18x20/#multiple-sensors

**Comment by lalo-uy** [comment]
With each branch having a separate Relay# as Vcc you can reset part of the
sensors when one gets stuck.
